code/sound/ds.cpp

```diff
@@ -306,12 +306,27 @@ int ds_load_buffer(int *sid, int  /*flags*/, sound::IAudioFile* file)
 }
 
 /**
- * Initialise the ::Channels[] array
+ * Initialise the ::Channels[] array dynamically based on system resources.
  */
 void ds_init_channels()
 {
-	MAX_CHANNELS = Cmdline_no_enhanced_sound ? 32 : 128;
-
+	// Legacy assumed safe value if for some reason dynamicly setting it does not work
+	MAX_CHANNELS = 128;
+
+	if (Cmdline_no_enhanced_sound)
+		MAX_CHANNELS = 32; // Old sound code limts
+	else {
+		ALCint size;
+		alcGetIntegerv(ds_sound_device, ALC_ATTRIBUTES_SIZE, 1, &size);
+		std::vector<ALCint> attrs(size);
+		alcGetIntegerv(ds_sound_device, ALC_ALL_ATTRIBUTES, size, &attrs[0]);
+		for (size_t i = 0; i < attrs.size(); ++i) {
+			if (attrs[i] == ALC_MONO_SOURCES) {
```

Comment:
As far as I know (my knowledge might be outdated) this doesn't necessarily report the number of playable sources. Last I checked the spec was unclear about what this actually reports and even then it can vary with the implementation. If OpenAL-Soft is required then that makes things easier, but it should be forced if that's the case. Otherwise you have to do a dynamic check to figure out how many can be played.

Reply:
This code was suggested by @TRBlount some time ago and I used it fairly uncritically. Don't we ship with our own openAL and can thus be sure of the implementation? Or is it deeper down than that?

If not, and this code is thus unsafe, my fallback would be to do somthing to the effect of `MAX_CHANNELS = attrs[i + 1]*0.9f;` (plus whatever casting is needed of course) to give a safety factor. The current 128 is a limit I surpass handily when I run tests with limits removed. It doesn't have noticable performance impact for me, it's possible that it will for others but this is a mod that is knowingly high-spec already.

If even that is not acceptable, I'll just drop this section of the changes from the PR, and possibly revisit it later in a separate one.

Reply:
It's definitely wrong with the way the code currently works (see my other comment/clarification) since it probably needs to be 30-40 less than that value for safety. But as long as OpenAL Soft is all we want to work with then the attributes method should work okay I'm guessing. Though it would be nice if, perhaps in another PR, we make that a point with a vendor string check combined with an error or stern warning. Assuming one doesn't already exist (haven't looked in a good number of years)

Since OpenAL Soft mixing is entirely CPU bound, performance is really going to vary based on the individual computer and perhaps the audio backend as well. It might be a good fallback to have a low-medium-high option for people that have sound related performance problems. With "high" being the max/default, "medium" being half that, and "low" being 32. Perhaps it could replace/augment the current cmdline option or something like that.

Reply:
I've pushed some changes to preserve an amount of headroom.